### PR TITLE
[WIP] Add unit tests.

### PR DESCRIPTION
**Pull request by petrabarus**
Still work in progress. Need to create the tests for important classes, and check for coverage for every unit tests in the tested class.

- [x] Unit test for ByteHelper
- [x] Unit test for S3Destination

### CODE REVIEW

File: src/main/java/tdl/s3/upload/MultipartUploadFileUploadingStrategy.java [hunk 1]
```diff
@@ -150,16 +158,23 @@ private void uploadRequiredParts(File file, String remotePath) throws IOExceptio
 
     private void submitUploadRequestStream(Stream<UploadPartRequest> requestStream) {
         requestStream.map(concurrentUploader::submitTaskForPartUploading)
-                .map(this::getUploadingResult)
+                .map(future -> {
+                    try {
+                        return this.getUploadingResult(future);
+                    } catch (DestinationOperationException ex) {
```
Comment (julianghionoiu) [inline src/main/java/tdl/s3/upload/MultipartUploadFileUploadingStrategy.java]: This is ok as failing to upload a part will not lead to problems.

File: src/main/java/tdl/s3/upload/MultipartUploadFileUploadingStrategy.java [hunk 2]
```diff
@@ -169,19 +184,25 @@ private UploadPartRequest getUploadPartRequest(String remotePath, byte[] nextPar
...
             throw new RuntimeException(ioe.getMessage(), ioe);
         }
     }
 
-    private MultipartUploadResult getUploadingResult(Future<MultipartUploadResult> future) {
+    private MultipartUploadResult getUploadingResult(Future<MultipartUploadResult> future) throws DestinationOperationException {
         try {
             return future.get();
-        } catch (InterruptedException | ExecutionException e) {
+        } catch (InterruptedException e) {
```
Comment (julianghionoiu) [inline src/main/java/tdl/s3/upload/MultipartUploadFileUploadingStrategy.java]: Very nice, I has not aware of the possibility of this failing.

File: src/main/java/tdl/s3/upload/FileUploaderImpl.java [hunk 3]
```diff
@@ -31,7 +34,11 @@ public void upload(File file, String path) {
 
     @Override
     public boolean exists(String path) {
-        return destination.canUpload(path);
+        try {
+            return destination.canUpload(path);
+        } catch (DestinationOperationException ex) {
+            return false;
```
Comment (petrabarus) [inline src/main/java/tdl/s3/upload/FileUploaderImpl.java]: What do you think of this?

https://github.com/julianghionoiu/s3-sync-stream/pull/26/commits/4042eb3fd8ee1f707bda0bc64014649994ed7884

File: src/main/java/tdl/s3/upload/MultipartUploadFileUploadingStrategy.java (hunk 1 above)
Comment (petrabarus) [inline src/main/java/tdl/s3/upload/MultipartUploadFileUploadingStrategy.java]: I guess we should log this as error even if it's gracefully returns null for failure.

Comment (julianghionoiu) [inline src/main/java/tdl/s3/upload/MultipartUploadFileUploadingStrategy.java]: Yes

File: src/main/java/tdl/s3/upload/FileUploaderImpl.java (hunk 3 above)
Comment (julianghionoiu) [inline src/main/java/tdl/s3/upload/FileUploaderImpl.java]: Yes, that I alright

File: src/main/java/tdl/s3/upload/MultipartUploadFileUploadingStrategy.java [hunk 4]
```diff
@@ -128,13 +131,18 @@ private void uploadRequiredParts(File file, String remotePath) throws IOExceptio
...
-    }
-
-    private Stream<UploadPartRequest> streamUploadForIncompleteParts(String remotePath, InputStream inputStream, boolean writingFinished) throws IOException {
+                    try {
+                        byte[] partData = ByteHelper.readPart(partNumber, file);
+                        UploadPartRequest request = getUploadPartRequest(remotePath, partData, false, partNumber);
+                        uploadedSize += partData.length;
+                        return request;
+                    } catch (IOException | DestinationOperationException ex) {
+                        return null;
```
Comment (petrabarus) [inline src/main/java/tdl/s3/upload/MultipartUploadFileUploadingStrategy.java]: Yes, now that you mentioned it. I think that maintaining a state `uploadedSize` in a process like this kind of defeat the purpose of functional programming.

**Comment by petrabarus** [comment]
Anyway just found out that `PerformanceTest.uploadPartialLargeMultipartFile` sometimes fails, I think due to connection.

File: src/main/java/tdl/s3/upload/MultipartUploadFileUploadingStrategy.java (hunk 1 above)
Comment (petrabarus) [inline src/main/java/tdl/s3/upload/MultipartUploadFileUploadingStrategy.java]: Done https://github.com/julianghionoiu/s3-sync-stream/pull/26/commits/29d055bdee9308c653b1b833e3940afa92fa0036

File: src/main/java/tdl/s3/helpers/ByteHelper.java [hunk 5]
```diff
@@ -43,20 +43,24 @@ public static ByteArrayInputStream createInputStream(byte[] bytes) {
...
     }
 
-    public static void skipOffsetInInputStream(InputStream stream, long offset) throws IOException {
+    static void skipOffsetInInputStream(InputStream stream, long offset) throws IOException {
         long skipped = 0;
-        while (skipped < offset) {
+        long trial = 0;
+        while (trial < 10) {
             skipped += stream.skip(offset);
+            if (skipped >= offset) {
```
Comment (petrabarus) [inline src/main/java/tdl/s3/helpers/ByteHelper.java]: Done https://github.com/julianghionoiu/s3-sync-stream/pull/26/commits/a439269d28d68ba1f4659d70776f90c88a530819

**Comment by petrabarus** [comment]
@julianghionoiu I mistakenly wrote an infinite loop. I saw the build hasn't ended yet. Can you cancel the build for https://github.com/julianghionoiu/s3-sync-stream/pull/26/commits/b47b9bd9a0dda141c5071e76f18ca7f04ee5f6e7 ?

**Comment by julianghionoiu** [comment]
I have stopped the build

**Pull request merged**